Update robust glm notebook

jonsedar opened the pull request:
The **first purpose of this PR** is to update the GLM robust regression notebook already in the examples here: https://docs.pymc.io/notebooks/GLM-robust-with-outlier-detection.html 

Those updates are everything from v2.1 onwards:

Version history:

version | date | author | changes
:--- | :--- | :--- | :---
1.0 | 2015-12-21 | [jonsedar](https://github.com/jonsedar) | Create and publish
2.0 | 2018-07-24 | [twiecki](https://github.com/twiecki) | Restate outlier model using `pm.Normal.dist().logp()` and `pm.Potential()`
2.1 | 2019-11-16 | [jonsedar](https://github.com/jonsedar) | Restate `nu` in StudentT model to be more efficient, drop explicit use of theano shared vars, generally improve plotting / explanations / layout
2.2 | 2020-05-21 | [jonsedar](https://github.com/jonsedar) | Minor tidyup for plots and warnings and rerun with pymc3.8


The **second purpose of this PR** is to clarify the docstring within `sampling.py`, specifically the kwargs for `step_kwargs` when you have multiple steppers. I found the need for better clarity in the docstring during my rework of the notebook, so I hope it's valid to include a change in this single PR. I think this is also a fix for https://github.com/pymc-devs/pymc3/issues/3197

Commit:
updated the Hogg notebook

Commit:
updated the Hogg notebook

Commit:
attempted to clarify the kwargs in sample() docstring describing how to pass kwargs to the steppers

I believe this fixes https://github.com/pymc-devs/pymc3/issues/3197

I also noted this need for more clarity in my updated notebook in this PR

`pymc3/docs/source/notebooks/GLM-robust-with-outlier-detection.ipynb`

Commit:
attempted to clarify the kwargs in sample() docstring describing how to pass kwargs to the steppers

I believe this fixes https://github.com/pymc-devs/pymc3/issues/3197

I also noted this need for more clarity in my updated notebook in this PR

`pymc3/docs/source/notebooks/GLM-robust-with-outlier-detection.ipynb`

michaelosthege committed:
Remove deprecated stuff (#3906)

* remove file which is not used

* remove deprecated code

* repair tests and notebooks that used deprecated API

* mention #3906

Co-authored-by: Michael Osthege <zufallsprinzip@hotmail.de>

michaelosthege committed:
add deprecation warnings for old backends (#3902)

* add deprecation warnings for old backends

* mention backend deprecation #3902

* fix typo

Co-authored-by: Colin <ColCarroll@users.noreply.github.com>

Co-authored-by: Michael Osthege <zufallsprinzip@hotmail.de>
Co-authored-by: Colin <ColCarroll@users.noreply.github.com>

twiecki commented:
Wow, this is an amazing improvement on an already amazing post.

The one thing I want to make sure is that the headings are correct -- are section headings starting with `##` rather than `#` which is reserved only for the title? Otherwise this creates problems with TOCs.

twiecki commented on pymc3/sampling.py:
```diff
@@ -334,8 +334,28 @@ def sample(
+    can correctly anticipate what they will be, then you can address them here
+    with keys that are (usually) the snakey_lowercase names of the step method.
+    e.g. to address:
+        1. ``target_accept`` to NUTS: nuts={'target_accept':0.9}
```
for nuts this also works if you pass it directly, right? This might make it sound like you always have to use this extra-nesting.

AlexAndorra replied:
Yes, it works if you pass it directly, i.e `pm.sample(target_accept=0.9)`. You have to pass the step method only when the sampler uses several step methods.
So, I agree with Thomas and think this clarification should be mentionned.

jonsedar replied:
Thanks for the feedback guys - I've clarified the language in my latest commit

Commit:
minor formatting to notebook and rework of docstring for sample function

notebook: dropped all headings one level lower to comply with TOC logic, and very minor language edits

sampling.py: clarifired language around single vs compoundstep

Commit:
minor formatting to notebook and rework of docstring for sample function

notebook: dropped all headings one level lower to comply with TOC logic, and very minor language edits

sampling.py: clarifired language around single vs compoundstep

jonsedar commented:
Thanks for the feedback guys - I've tidied the notebook and clarified the docstring language in my latest commit

AlexAndorra commented:
Thanks Jon! I'm reviewing the NB right now -- will let you know when I'm done

AlexAndorra commented:
Thanks Jon, this looks really nice and is a clear improvement over the old NB! I posted the first part of my review of your NB above, and I'll review the second part later this afternoon 😉

jonsedar commented:
Thanks Alex, I'll wait for you to complete the review before I make any changes that you've suggested.

Cheers,

AlexAndorra commented:
Yeah that'd be best if you can. I'm almost finished.

AlexAndorra reviewed (changes requested):
The docstring looks good now 👌 I had a few comments for the NB -- should be ready to merge after that.
This is a great update, thanks Jon!

jonsedar commented:
Righto - thanks for the thorough review! I'll make changes after lunch :)

jonsedar commented:
<p>Yep - removed. The important bits (python 3.6, pymc3 3.8) are already noted</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3908/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jonsedar commented:
<p>probably worth keeping incase people run this in other IDEs (VSCode, PyCharm, Syder etc) I'm not sure how those work</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3908/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jonsedar commented:
<p>Sure, done.</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3908/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jonsedar commented:
<p>Yeah, that can go - I tend to keep this around in my WIP templates as a location to abstract away long functions before they go to a source file</p><p> </p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3908/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jonsedar commented:
<p>ta :) though sometimes I really miss ggplot text_geom</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3908/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jonsedar commented:
<p>good catch</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3908/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jonsedar commented:
<p>Cool, will make a mental note to update this once that's out :)</p><p> </p><p>I kept the defaults in since I think it can be useful to see them</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3908/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jonsedar commented:
<p>Good to know - I'll definitely look out for that in future then: I find joint dists are massively helpful</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3908/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jonsedar commented:
<p>Gotcha :)</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3908/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jonsedar commented:
<p>Thanks, I've made a note to come back update arvin throughout when available https://github.com/jonsedar/pymc3_examples/issues/15</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3908/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jonsedar commented:
<p>good catch</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3908/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jonsedar commented:
<p>Fair point, added a note:</p><p>testval for <code>is_outlier</code> initialised in order to create class asymmetry</p><p> </p><p>The other testvals <em>were</em> set also in order to create asymmetry, but I've found they don't need it. Have reverted to mean values for b0, b1, y_est_out, sigma_y_out</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3908/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jonsedar commented:
<p>The passing of time eh? have updated to arviz=0.7 released between initial draft writing and now. I've been able to remove the warning suppression</p><p> </p><p>Will remove the <code>combined</code> and <code>compact</code>   kwargs: I don't think they used to be false by default. All good</p><p> </p><p>That's a nifty trick with the <code>var_names</code>,  will remember that for later. I think for now though it's safer to make explicit wouldn't you say?</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3908/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jonsedar commented:
<p>Outliers ought to be the minority class (and might not even exist) in this modelspec, so I think it's fair to constrain them Uniform[0, 0.5], rather than Beta(0, 1)...</p><p> </p><p>But, that said, Uniform is a bit of a cludge. Could possibly use a distribution with a more central tendency, but would have to truncate it. Maybe there's another modelspec somewhere :)</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3908/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jonsedar commented:
<p>changed to az </p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3908/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jonsedar commented:
<p>righto - I've logged a note to update these at some point :) https://github.com/jonsedar/pymc3_examples/issues/15</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3908/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jonsedar commented:
<p>Added to the list :) https://github.com/jonsedar/pymc3_examples/issues/15</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3908/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jonsedar commented:
<p>ah sweet typos... good catch, thanks!</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3908/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

Commit:
Merge branch 'update-robust-glm-notebook' of https://github.com/jonsedar/pymc3 into update-robust-glm-notebook

Commit:
updates folowing AlexAndorra review:

upgrade to arviz=0.7
set prior params to slightly simpler (more justifiable) values, and testvals to simplier defaults
explanatory clarifications
formatting, typos,

jonsedar commented:
okie dokie, I've addressed all @AlexAndorra's points - great review thank you - and I think this is ready to go

One note: I did the olde:

```
❯ git fetch upstream
❯ git rebase upstream/master
```

prior to my latest commit in this PR and see quite a few files changed in the meantime , must be a weekend with lots of folks working :)

I don't see any clashes, but do let me know if you'd prefer a clean PR

Cheers, Jon

Commit:
removed the note re step_kwargs, since this PR updates the appropriate docstring

fonnesbeck commented:
Hey, Jon. Looks great. Why is section 4.2.3 a markdown cell and not code?

AlexAndorra commented:
Thanks for the updates Jon! I'll review that tomorrow

Le dim. 3 mai 2020 à 21:20, Jonathan Sedar <notifications@github.com> a
écrit :

> okie dokie, I've addressed all @AlexAndorra
> <https://github.com/AlexAndorra>'s points - great review thank you - and
> I think this is ready to go
>
> One note: I did the olde:
>
> ❯ git fetch upstream
>
> ❯ git rebase upstream/master
>
>
> prior to my latest commit in this PR and see quite a few files changed in
> the meantime , must be a weekend with lots of folks working :)
>
> I don't see any clashes, but do let me know if you'd prefer a clean PR
>
> Cheers, Jon
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pymc-devs/pymc3/pull/3908#issuecomment-623165944>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHIJMTFA5ISJDKYS4S5MVTDRPW7XFANCNFSM4MX4IOIA>
> .
>

jonsedar commented:
> Hey, Jon. Looks great. Why is section 4.2.3 a markdown cell and not code?

Quite right - I'm an idiot! Will update now

Commit:
a cell had become markdown, silly. reset it to code and rerun

AlexAndorra reviewed (changes requested):
Thanks for the updates Jon, this all looks good now! Just one last thing (sorry, forgot to tell you about it yesterday): could you please Black-format the NB with [black_nbconvert](https://github.com/dfm/black_nbconvert)? 
That's what we use on the resources repo to standardize NBs: `pip install black_nbconvert`, then `black_nbconvert /path/to/a/notebook.ipynb`. 
After that I'll merge 🍾 

A lot of files have indeed been picked up by rebasing your branch. I don't know how you guys usually do, but I think it's no prb, as there is no conflicts and I'll squash & merge anyway.

jonsedar commented:
Hi Alex,

Good call re: https://github.com/dfm/black_nbconvert, that's a new one to me, ands no surprise to see it's by the ever-helpful & prolific DanFM!

I've just run that reformatting and a re-run, all seems well, so will commit now

Commit:
minor code reformatting via black_nbconvert, final check and re-run

jonsedar commented:
Interesting test failure, if this requires underlying fixes to pymc3 let me know and perhaps for simplicity I can just make a new clean PR

AlexAndorra reviewed (approved):
Thanks a lot Jon, this is all good now 👌 
I'll merge as soon as tests pass -- this error is super weird BTW. It looks like pip couldn't install numpy. Do you have an idea how to fix that?

jonsedar commented:
No idea - it could just have been a network hiccup at the time... For good measure I'll recommit and trigger again!

Commit:
rerun notebook purely as a lazy but safe way to trigger new CI

jonsedar commented:
It was a ghost in the machine :D

AlexAndorra commented:
Apparently 😅 
Thanks again for this great update Jon!

jonsedar commented:
Thanks for all your help and reviews! Will try to pick up some bugs next...

BTW is there a protocol for deleting this branch on origin?

AlexAndorra commented:
Good question -- I'm not aware of any such protocol, but that'd be good practice
